File: sorting/run_kilosort4.m
function run_kilosort4(cfg, day, override, rec, tower, np, grouped)
% RUN_KILOSORT4 Execute Kilosort4 spike sorting on SpikeGLX recordings
%
%   run_kilosort4(cfg, day, override, rec, tower, np, grouped)
%
%   Concatenates SpikeGLX .bin files (if grouped), then runs Kilosort4
%   via a Python virtualenv to produce spike sorting results.
%
%   Parameters:
%     cfg      - Configuration struct from mind_snag_config
%     day      - Recording date as string, YYMMDD format (e.g. '250224')
%     override - Logical. If true, re-concatenate even if combined file exists
%     rec      - Cell array of recording numbers (e.g. {'007','009','010'})
%     tower    - Recording setup name (e.g. 'LPPC_LPFC_modularV1')
%     np       - Neuropixel probe number (1 or 2)
%     grouped  - 0 for individual recordings, 1 for concatenated
%
%   Requires:
%     - Python virtualenv with kilosort[gui]==4.0.22 at cfg.kilosort_venv
%     - Probe file at cfg.probe_file
%     - Run_kilosort4.py at cfg.kilosort_script
%
%   See also: extract_spikes, mind_snag_config

% Validate inputs
assert(~isempty(cfg.data_root), 'mind_snag:config', ...
    'cfg.data_root must be set. Run mind_snag_config first.');
assert(~isempty(cfg.kilosort_venv), 'mind_snag:config', ...
    'cfg.kilosort_venv must be set to your Python virtualenv path.');
assert(exist(cfg.data_root, 'dir') == 7, 'mind_snag:io', ...
    'Data root not found: %s', cfg.data_root);

data_root = cfg.data_root;
gpu = cfg.gpu;

grouped_recs{1} = rec;
ngroups = length(grouped_recs);

%% Create grouped recordings directory
group_rec_dir = fullfile(data_root, day, 'spikeglx_data', ...
    sprintf('grouped_recordings.%s.%d', tower, np));

if ~exist(group_rec_dir, 'dir')
    mkdir(group_rec_dir);
end

fprintf('Saving spike sorting rec groups\n');
if grouped
    grouped_name = strjoin(rec, '_');
    save(fullfile(group_rec_dir, ...
        ['spike_sorting_rec_groups_' grouped_name '.mat']), ...
        'grouped_recs');
else
    grouped_name = rec{1};
    save(fullfile(group_rec_dir, ...
        ['spike_sorting_rec_groups_' rec{1} '.mat']), 'grouped_recs');
end

%% Concatenate input data
for igroup = 1:ngroups
    group = grouped_recs{igroup};
    if grouped == 0
        group = {group};
    end
    imec_rec_dirs = {};
    sglx_files_to_join = {};
    ap_metas = {};

    nsamp_total = 0;

    for iR = 1:length(group)
        ap_meta_file = fullfile(data_root, day, rec{iR}, ...
            sprintf('rec%s.%s.%d.ap_meta.mat', rec{iR}, tower, np));
        assert(exist(ap_meta_file, 'file') == 2, 'mind_snag:io', ...
            'AP meta file not found: %s', ap_meta_file);

        load(ap_meta_file, 'ap_meta');

        ap_metas{iR} = ap_meta;
        nsamp_total = nsamp_total + ap_meta.nsamp;
        [pathstr, ~, ~] = fileparts(ap_meta.fileName);
        dirs = strsplit(pathstr, '/');
        match_idx = find(~cellfun('isempty', regexp(dirs, 'rec')));
        sglx_rec = dirs{match_idx};

        if isfield(ap_meta, 'imec_used_probe_name')
            imec_rec_dir = fullfile(data_root, day, ...
                'spikeglx_data', sglx_rec, ...
                [sglx_rec, sprintf('_%s', ap_meta.imec_used_probe_name)]);
            ap_bin_name = [sglx_rec, sprintf('_t0.%s.ap.bin', ...
                ap_meta.imec_used_probe_name)];
        elseif isfield(ap_meta, 'used_probe_name')
            imec_rec_dir = fullfile(data_root, day, ...
                'spikeglx_data', sglx_rec, ...
                [sglx_rec, sprintf('_%s', ap_meta.used_probe_name)]);
            ap_bin_name = [sglx_rec, sprintf('_t0.%s.ap.bin', ...
                ap_meta.used_probe_name)];
        else
            warning('mind_snag:probe', 'Probe name missing; assuming imec0');
            ap_meta.imec_used_probe_name = 'imec0';
            imec_rec_dir = fullfile(data_root, day, ...
                'spikeglx_data', sglx_rec, ...
                [sglx_rec, sprintf('_%s', ap_meta.imec_used_probe_name)]);
            ap_bin_name = [sglx_rec, sprintf('_t0.%s.ap.bin', ...
                ap_meta.imec_used_probe_name)];
        end

        if contains(ap_bin_name, 'reco')
            ap_bin_name = strrep(ap_bin_name, 'reco', 'rec');
        end
        if contains(imec_rec_dir, 'reco')
            imec_rec_dir = strrep(imec_rec_dir, 'reco', 'rec');
        end

        imec_rec_dirs{iR} = imec_rec_dir;
        sglx_files_to_join{iR} = ap_bin_name;
    end

    nchan = ap_meta.nchan;

    %% Copy and merge files
    group_rec_input_dir = fullfile(data_root, day, 'spikeglx_data', ...
        sprintf('rec%s', grouped_name), ...
        sprintf('rec%s_imec%d', grouped_name, (np-1)));
    if ~exist(group_rec_input_dir, 'dir')
        mkdir(group_rec_input_dir);
    end

    combined_ap_bin = fullfile(group_rec_input_dir, ...
        sprintf('combined_ap_group%s.bin', grouped_name));

    expected_bytes = nsamp_total * nchan * 2;

    combined_ap_exists = false;
    if exist(combined_ap_bin, 'file')
        file_info = dir(combined_ap_bin);
        if file_info.bytes == expected_bytes
            combined_ap_exists = true;
            fprintf('  Combined AP file already written\n');
        end
    end

    if ~combined_ap_exists || override
        fid = fopen(combined_ap_bin, 'w');
        fclose(fid);
        fid = fopen(combined_ap_bin, 'a');

        for iR = 1:length(group)
            fprintf('  Memory mapping rec %s\n', rec{iR});
            m_ap = sglx_util.memmap(ap_metas{iR}, sglx_files_to_join{iR}, ...
                imec_rec_dirs{iR});
            nsamp = size(m_ap.Data.x, 2);
            blocksize = 500;
            nblocks = ceil(nsamp / blocksize);
            fprintf('  Writing data\n');
            for iblock = 1:nblocks
                idx = (iblock-1)*blocksize + 1 : min(nsamp, iblock*blocksize);
                fwrite(fid, m_ap.Data.x(:, idx), 'int16');
            end
        end
        fclose(fid);
    end
end

%% Generate Kilosort4 output folder
if grouped == 1
    KS_result_dir = fullfile(group_rec_dir, ['group', grouped_name '_KS4']);
else
    KS_result_dir = fullfile(group_rec_dir, ['group', rec{1} '_KS4']);
end
if ~exist(KS_result_dir, 'dir')
    mkdir(KS_result_dir);
end

%% Determine input data
if grouped == 1
    fprintf('Input: concatenated recordings %s\n', grouped_name);
    InputData = combined_ap_bin;
else
    recordings_excel_file = fullfile(data_root, 'excel', 'valid_recordings.xlsx');
    assert(exist(recordings_excel_file, 'file') == 2, 'mind_snag:io', ...
        'Recording metadata not found: %s', recordings_excel_file);
    all_records = excel_table_to_struct_array(recordings_excel_file);
    matching_indices = strcmp({all_records.Day}, day);
    records_for_specific_day = all_records(matching_indices);
    matching_indices_Rec = strcmp({records_for_specific_day.Rec}, rec);
    records_for_specific_day_rec = records_for_specific_day(matching_indices_Rec);
    Sglx = records_for_specific_day_rec.SglxRec;
    InputData = fullfile(data_root, day, 'spikeglx_data', Sglx, ...
        [Sglx, '_imec', num2str(np-1)], ...
        [Sglx, '_t0.imec', num2str(np-1), '.ap.bin']);
end

resultsDir = KS_result_dir;
fprintf('Input: %s\n', InputData);
fprintf('Output: %s\n', resultsDir);

assert(exist(InputData, 'file') == 2, 'mind_snag:io', ...
    'Input binary not found: %s', InputData);

%% Run Kilosort4 via Python
scriptName = cfg.kilosort_script;
probeFile = cfg.probe_file;
paramsDir = cfg.ks_params_dir;
venvPath = cfg.kilosort_venv;

assert(exist(venvPath, 'dir') == 7, 'mind_snag:config', ...
    'Python virtualenv not found: %s', venvPath);

fprintf('Starting Kilosort4...\n');
command = sprintf(['bash -lc "source %s/bin/activate && ' ...
    'export OPENBLAS_NUM_THREADS=%d && ' ...
    'python \\"%s\\" --results \\"%s\\" --probe \\"%s\\" ' ...
    '--params \\"%s\\" --data \\"%s\\" --gpu %d"'], ...
    venvPath, cfg.n_threads, scriptName, resultsDir, probeFile, ...
    paramsDir, InputData, gpu);

[status, cmdOut] = system(command);

if status ~= 0
    warning('mind_snag:kilosort', 'Kilosort4 exited with status %d:\n%s', status, cmdOut);
else
    fprintf('Kilosort4 completed successfully.\n');
end

end

File: mind_snag_py/src/mind_snag/sorting/run_kilosort4.py
"""Execute Kilosort4 spike sorting on SpikeGLX recordings.

Ports run_kilosort4.m. Calls the kilosort Python API directly instead of
shelling out to a subprocess. Handles .bin file concatenation for grouped mode.
"""

from __future__ import annotations

import logging
from pathlib import Path

import numpy as np

from mind_snag.config import MindSnagConfig
from mind_snag.io.mat_reader import load_mat
from mind_snag.utils.paths import group_rec_dir, rec_name_str

logger = logging.getLogger(__name__)


def run_kilosort4(
    cfg: MindSnagConfig,
    day: str,
    recs: list[str],
    tower: str,
    np_num: int,
    grouped: bool,
    override: bool = False,
) -> Path:
    """Execute Kilosort4 on SpikeGLX recordings.

    Parameters
    ----------
    cfg : pipeline configuration
    day : recording date (YYMMDD)
    recs : list of recording numbers
    tower : recording setup name
    np_num : probe number
    grouped : whether to concatenate recordings
    override : if True, re-concatenate even if combined file exists

    Returns
    -------
    Path to KS4 results directory
    """
    cfg.validate()
    data_root = cfg.data_root
    rec_str = rec_name_str(recs, grouped)

    # Create grouped recordings directory
    grp_dir = group_rec_dir(data_root, day, tower, np_num)
    grp_dir.mkdir(parents=True, exist_ok=True)

    # Concatenate input data if grouped
    if grouped:
        combined_bin = _concatenate_bins(
            cfg, day, recs, tower, np_num, rec_str, grp_dir, override,
        )
        input_data = combined_bin
    else:
        input_data = _find_single_rec_bin(cfg, day, recs[0], tower, np_num)

    # Create KS4 output directory
    ks_result_dir = grp_dir / f"group{rec_str}_KS4"
    ks_result_dir.mkdir(parents=True, exist_ok=True)

    logger.info("Input: %s", input_data)
    logger.info("Output: %s", ks_result_dir)

    if not input_data.exists():
        raise FileNotFoundError(f"Input binary not found: {input_data}")

    # Run Kilosort4 via Python API
    _run_ks4_api(input_data, ks_result_dir, cfg)

    return ks_result_dir


def _concatenate_bins(
    cfg: MindSnagConfig,
    day: str,
    recs: list[str],
    tower: str,
    np_num: int,
    rec_str: str,
    grp_dir: Path,
    override: bool,
) -> Path:
    """Concatenate .ap.bin files from multiple recordings.

    Uses numpy memory mapping instead of MATLAB's sglx_util.memmap.
    """
    data_root = cfg.data_root

    ap_metas = []
    bin_files = []
    nsamp_total = 0

    for rec in recs:
        ap_meta_file = data_root / day / rec / f"rec{rec}.{tower}.{np_num}.ap_meta.mat"
        if not ap_meta_file.exists():
            raise FileNotFoundError(f"AP meta file not found: {ap_meta_file}")

        meta_data = load_mat(ap_meta_file)
        ap_meta = meta_data.get("ap_meta", {})
        ap_metas.append(ap_meta)

        nsamp = int(ap_meta.get("nsamp", 0)) if isinstance(ap_meta, dict) else int(getattr(ap_meta, "nsamp", 0))
        nsamp_total += nsamp

        # Find the .ap.bin file
        bin_path = _find_rec_bin(data_root, day, rec, ap_meta, np_num)
        bin_files.append(bin_path)

    # Get nchan from last meta
    nchan = int(ap_meta.get("nchan", 385)) if isinstance(ap_meta, dict) else int(getattr(ap_meta, "nchan", 385))

    # Output path
    group_input_dir = data_root / day / "spikeglx_data" / f"rec{rec_str}" / f"rec{rec_str}_imec{np_num - 1}"
    group_input_dir.mkdir(parents=True, exist_ok=True)
    combined_bin = group_input_dir / f"combined_ap_group{rec_str}.bin"

    expected_bytes = nsamp_total * nchan * 2  # int16

    if combined_bin.exists() and combined_bin.stat().st_size == expected_bytes and not override:
        logger.info("Combined AP file already exists: %s", combined_bin)
        return combined_bin

    logger.info("Concatenating %d recordings into %s", len(recs), combined_bin)

    with open(combined_bin, "wb") as fout:
        for i, (rec, bin_path) in enumerate(zip(recs, bin_files)):
            if not bin_path.exists():
                raise FileNotFoundError(f"Binary file not found: {bin_path}")
            meta = ap_metas[i]
            nsamp = int(meta.get("nsamp", 0)) if isinstance(meta, dict) else int(getattr(meta, "nsamp", 0))
            logger.info("  Memory mapping rec %s (%d samples)", rec, nsamp)

            # Memory-map and write in blocks
            mmap = np.memmap(bin_path, dtype=np.int16, mode="r",
                             shape=(nsamp, nchan))
            block_size = 500
            n_blocks = (nsamp + block_size - 1) // block_size
            for ib in range(n_blocks):
                start = ib * block_size
                end = min(nsamp, (ib + 1) * block_size)
                fout.write(mmap[start:end].tobytes())
            del mmap

    return combined_bin


def _find_rec_bin(
    data_root: Path, day: str, rec: str, ap_meta: dict, np_num: int,
) -> Path:
    """Find the .ap.bin file for a recording from its metadata."""
    if isinstance(ap_meta, dict):
        probe_name = ap_meta.get("imec_used_probe_name") or ap_meta.get("used_probe_name") or f"imec{np_num - 1}"
        file_name = ap_meta.get("fileName", "")
    else:
        probe_name = getattr(ap_meta, "imec_used_probe_name", None) or getattr(ap_meta, "used_probe_name", None) or f"imec{np_num - 1}"
        file_name = getattr(ap_meta, "fileName", "")

    # Try to find in spikeglx_data directory
    sglx_dir = data_root / day / "spikeglx_data"
    # Look for patterns like rec007/rec007_imec0/rec007_t0.imec0.ap.bin
    for rec_dir in sglx_dir.glob(f"*{rec}*"):
        if rec_dir.is_dir():
            for imec_dir in rec_dir.glob(f"*{probe_name}*"):
                if imec_dir.is_dir():
                    for bin_file in imec_dir.glob("*.ap.bin"):
                        return bin_file

    # Fallback: construct path directly
    sglx_rec = f"rec{rec}"
    imec_rec_dir = sglx_dir / sglx_rec / f"{sglx_rec}_{probe_name}"
    bin_name = f"{sglx_rec}_t0.{probe_name}.ap.bin"
    return imec_rec_dir / bin_name


def _find_single_rec_bin(
    cfg: MindSnagConfig, day: str, rec: str, tower: str, np_num: int,
) -> Path:
    """Find the .ap.bin file for a single (non-grouped) recording."""
    data_root = cfg.data_root
    # Try loading AP meta to find path
    ap_meta_file = data_root / day / rec / f"rec{rec}.{tower}.{np_num}.ap_meta.mat"
    if ap_meta_file.exists():
        meta_data = load_mat(ap_meta_file)
        ap_meta = meta_data.get("ap_meta", {})
        return _find_rec_bin(data_root, day, rec, ap_meta, np_num)

    # Fallback: look in spikeglx_data
    sglx_dir = data_root / day / "spikeglx_data"
    for bin_file in sglx_dir.rglob(f"*{rec}*.ap.bin"):
        return bin_file

    raise FileNotFoundError(
        f"Could not find .ap.bin for rec {rec}, day {day}"
    )


def _run_ks4_api(
    input_data: Path,
    results_dir: Path,
    cfg: MindSnagConfig,
) -> None:
    """Run Kilosort4 via the Python API.

    Uses ``kilosort.run_kilosort()`` directly instead of subprocess.
    """
    try:
        from kilosort import run_kilosort
        from kilosort.parameters import DEFAULT_SETTINGS
    except ImportError as e:
        raise ImportError(
            "kilosort package not installed. Install with: pip install 'kilosort>=4.0.20'"
        ) from e

    settings = dict(DEFAULT_SETTINGS)
    settings["data_dir"] = str(input_data.parent)
    settings["results_dir"] = str(results_dir)
    settings["n_chan_bin"] = 385  # default for NP1
    settings["batch_size"] = 60000

    logger.info("Starting Kilosort4...")
    run_kilosort(
        settings=settings,
        filename=str(input_data),
        results_dir=results_dir,
    )
    logger.info("Kilosort4 completed successfully.")
